### ISSUE TITLE
Use regex expression filters

### ISSUE THREAD

**Issue by richyeiv** [nodkz/graphql-compose#3]
In the following example:
```javascript
ViewerTC.addField('dmcList', DmcTC.getResolver('findMany').getFieldConfig());
```
I want to add a field 'dmcSearch' to filter using a regex expression.
Where can I start? I have been trying with something like:
```javascript
ViewerTC.addField('dmcSearch', {
  type: DmcTC.getType(),
  args: {
    expression: { type: GraphQLString }
  },
  resolve: (dmc, {expression}) => {
    
  }
});
```
Thanks in advance.

**Comment by nodkz** [comment]
Right now it is possible but too complicated:

```js 
ViewerTC.addField(
  'dmcList', 
  DmcTC
    .getResolver('findMany')
    // wrap resolver to extend filtering: 
    //   for removing `expression` from mongoose filter  
    //   for adding `MONGO_FIELD_NAME` to mongooseModel.where({ ..., MONGO_FIELD_NAME: regex });
    .wrapResolve(next => resolveParams => {
      if (resolveParams.args.filter && resolveParams.args.filter.expression) {
        const r = new RegExp(resolveParams.args.filter.expression, 'i');
        delete resolveParams.args.filter.expression;
        resolveParams.args.filter.MONGO_FIELD_NAME = r;
      }
      return next(resolveParams);
    })
    .getFieldConfig()
);

// Add `expression` arg to the `filter`
ViewerTC.getByPath('dmcList.@input.filter').addFields({
  expression: {
    type: GraphQLString,
  },
});
```

Try this solution, ping me if it works or not.


Your issue inspires me to simplify this process and create a declarative method for adding a filter to the resolver (cause it's the common task for graphql-composer).

I'll prepare a proposal for this new method and show you for review ASAP.

**Comment by richyeiv** [comment]
Thanks  for the answer, I try the following because 'name' is the property I want to filter by:
```javascript
ViewerTC.addField(
  'dmcList', 
  DmcTC
    .getResolver('findMany')
    .wrapResolve(next => resolveParams => {
      if (resolveParams.args.filter && resolveParams.args.filter.expression) {
        const r = new RegExp(resolveParams.args.filter.expression, 'i');
        delete resolveParams.args.filter.expression;
        resolveParams.args.filter.name = r;
      }
      return next(resolveParams);
    })
    .getFieldConfig()
);

ViewerTC.getByPath('dmcList.@input.filter').addFields({
  expression: {
    type: GraphQLString,
  },
});
```
But it returns an error:
```
ViewerTC.getByPath('dmcList.@input.filter').addFields({

TypeError: Cannot read property 'addFields' of undefined
```

**Comment by nodkz** [comment]
Oh, if you don't use Relay wrapper, then your path should be:
`ViewerTC.getByPath('dmcList.@filter')`

`@` - means find arg with name `filter`.

`dmcList.@input.filter` = find field `dmcList`, in its type find arg with name`input`, in argument find sub-field with name `filter`.

`dmcList.@input.filter` = find field `dmcList`, in this field find arg `filter`.

**Comment by richyeiv** [comment]
Hi, in fact I'm ussing relay wrapper, as I understand:
```javascript
const DmcTC = composeWithRelay(mongooseToTypeComposer(Dmc, customizationOptions));
````
where Dmc is a mongoose model.

But this still doesn't work and returns undefined:
```javascript
console.log(ViewerTC.getByPath('dmcList.@filter'));
```
Where ViewerTC is:
```javascript
composeWithRelay(GQC.rootQuery());
const ViewerTC = GQC.get('Viewer');
GQC.rootQuery().addField('viewer', {
  type: ViewerTC.getType(),
  description: 'Data under client context',
  resolve: () => ({}),
});
```

**Comment by nodkz** [comment]
So `undefined` returned if by `this` path is nothing exists.
Can you provide screenshots from graphiql of Viewer type.
I suppose that field `dmcList ` does not have any args or has different name.

**Comment by richyeiv** [comment]
Sure, here's the viewer, the dmclist and the filter:
![viewer](https://cloud.githubusercontent.com/assets/1832011/18361398/a0435602-75cf-11e6-881c-f5e1161b386b.PNG)
![dmclist](https://cloud.githubusercontent.com/assets/1832011/18361403/a2ae8ae2-75cf-11e6-84db-f8b4c21368b8.PNG)
![filterfindmanydmcinput](https://cloud.githubusercontent.com/assets/1832011/18361408/a7c2c796-75cf-11e6-9dd8-251f4913d2b1.PNG)

also I tryied the following code, but it returns all the items, no filter evaluated:
```javascript
var expressionInput = new GraphQLInputObjectType({
  name: 'regex',
  fields: {
    expression: { type: GraphQLString }
  }
});

var regexFieldConfig = DmcTC
    .getResolver('findMany')
    .wrapResolve(next => resolveParams => {
      if (resolveParams.args.filter && resolveParams.args.filter.expression) {
        const r = new RegExp(resolveParams.args.filter.expression, 'i');
        delete resolveParams.args.filter.expression;
        resolveParams.args.filter.name = r;
      }
      return next(resolveParams);
    })
    .getFieldConfig();
    
regexFieldConfig.args.filter.type = expressionInput;

ViewerTC.addField(
  'dmcList', 
  regexFieldConfig
);
```

**Comment by nodkz** [comment]
```js
ViewerTC.getByPath('dmcList.@filter').addFields({
  expression: {
    type: GraphQLString,
  },
});
```
According to your screenshots code above should work. It adds `expression` field to FilterFindManyDmcInput

**Comment by richyeiv** [comment]
Outputs the same error:
```
ViewerTC.getByPath('dmcList.@filter').addFields({
TypeError: Cannot read property 'addFields' of undefined
```
Maybe I am missing something. In the following code: 
`resolveParams.args.filter.MONGO_FIELD_NAME = r;`
MONGO_FIELD_NAME must be set to the field i want to query, right? In my case 'name'.

**Comment by nodkz** [comment]
Hmmm, too strange.

Ok try again...

**Step 1: Extending resolve method** of `dmcList` field.
All args that exist in `args.filter` will be implicitly passed to the `MongooseQuery.where` method.
So you may change `args.filter` as you wish for obtaining proper mongoose query to fit your business logic. For your case, we removing `regex` from the query and add `name`.

**Step 1 goal**: make working mongoose query
```js
ViewerTC.addField(
  'dmcList', 
  DmcTC
    .wrapResolve(next => resolveParams => {  // Step 1 extending resolve method
      if (resolveParams.args.filter && resolveParams.args.filter.regex) {
        const r = new RegExp(resolveParams.args.filter.regex, 'i');
        delete resolveParams.args.filter.regex;
        resolveParams.args.filter.name = r;
      }
      return next(resolveParams);
    })
    .getFieldConfig()
);
```

--
Be sure that `Step 1` **runs before** `Step 2`, because on the first step we also adding `dmcList` to the `Viewer` type. Otherwise `ViewerTC.getByPath('dmcList.@filter')` will return `undefined`.

**Step 2: adding field to InputType** in args
`dmcList` has args, one of them is `filter`. So we adding field to it with name `regex`.

**Step 2 goal**: See in GraphiQL for type `FilterFindManyDmcInput` new field `regex: String`. 
```js
ViewerTC.getByPath('dmcList.@filter').addFields({
  regex: {
    type: GraphQLString,
  },
});
```

So now we may send this query to GraphQL:
```js
{
  Viewer {
    dmcList(filter: { regex: "search text" }) {
      name
      status
      color
    }
  }
}
```

**Comment by richyeiv** [comment]
Hi again, it still returns the error described before. Maybe if you can provide a working example. If not we can close this. When you think you will have implemented the regex has an option in compose?
Thanks for all.

**Comment by richyeiv** [comment]
Hmm wait, I see something interesting in the output I will check in the following minutes.

**Comment by richyeiv** [comment]
With the last provided code, it outputs this:
```
ViewerTC.addField('dmcList', DmcTC.wrapResolve(function (next) {
TypeError: DmcTC.wrapResolve is not a function
```

**Comment by nodkz** [comment]
Should be `DmcTC.getResolver('findMany').wrapResolve`. Fix message above.

**Comment by richyeiv** [comment]
With that code, outputs the previously described error.

**Comment by nodkz** [comment]
On next week add an example to [relay-northwind-app](https://github.com/nodkz/relay-northwind-app).

**Comment by nodkz** [comment]
Closed due new `filter args proposal` in #22

**Comment by nodkz** [comment]
In `graphql-compose@0.1.1` and `graphql-compose-mongoose@1.2.1` added helper `Resolver.addFilterArg` for creation a custom filtering logic.

Code example with RegExp:
https://github.com/nodkz/graphql-compose-examples/blob/master/examples/northwind/models/product.js#L38,L49

Live demo: 
http://graphql-compose.herokuapp.com/northwind/?query=%0A%7B%0A%20%20viewer%20%7B%0A%20%20%20%20productList(filter%3A%20%7B%20nameRegexp%3A%20%22of%22%20%7D)%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%20%20%20%20%20%20
